src/test/java/seedu/address/model/session/RecurringSessionTest.java

```diff
@@ -1,41 +1,121 @@
+        if (CARL.getListOfSessions().get(0) instanceof RecurringSession) {
+            assertEquals(INTERVAL, (
+                    (RecurringSession) CARL.getListOfSessions().get(0)).getInterval());
+        }
     }
 
     @Test
-    void numOfSessionBetween() {
+    void getLastSessionDateTest() {
+        if (CARL.getListOfSessions().get(0) instanceof RecurringSession) {
+            assertEquals(new SessionDate("2021-01-15", DEFAULT_TIME), ((
+                    (RecurringSession) CARL.getListOfSessions().get(0)).getLastSessionDate()));
+        }
+    }
+
+    @Test
+    void hasSessionOnDateTest() {
+        RecurringSession startAfter = new RecurringSession(new SessionDate("2021-03-01", "10:00"),
+                DURATION, SUBJECT, FEE, INTERVAL, new SessionDate("2021-03-15", "10:00"));
+        RecurringSession endBefore = new RecurringSession(new SessionDate("2020-03-01", "10:00"),
+                DURATION, SUBJECT, FEE, INTERVAL, new SessionDate("2020-03-15", "10:00"));
+        RecurringSession notOnDate = new RecurringSession(new SessionDate("2020-12-31", "10:00"),
+                DURATION, SUBJECT, FEE, INTERVAL, new SessionDate("2021-01-14", "10:00"));
+        RecurringSession onDate = new RecurringSession(new SessionDate("2020-12-11", "10:00"),
+                DURATION, SUBJECT, FEE, INTERVAL, new SessionDate("2021-01-15", "10:00"));
+
+        assertFalse(startAfter.hasSessionOnDate(SESSION_DATE));
+        assertFalse(endBefore.hasSessionOnDate(SESSION_DATE));
+        assertFalse(notOnDate.hasSessionOnDate(SESSION_DATE));
+        assertTrue(onDate.hasSessionOnDate(SESSION_DATE));
+    }
+
+    @Test
+    void lastSessionOnOrBeforeTest() {
+        RecurringSession recurringSession = new RecurringSession(new SessionDate("2020-12-11", "10:00"),
+                DURATION, SUBJECT, FEE, INTERVAL, new SessionDate("2021-01-15", "10:00"));
+        Session session = new Session(new SessionDate("2021-01-01", "10:00"), DURATION, SUBJECT, FEE);
+        assertEquals(session, recurringSession.lastSessionOnOrBefore(SESSION_DATE));
+
+        RecurringSession startAfter = new RecurringSession(new SessionDate("2021-12-11", "10:00"),
+                DURATION, SUBJECT, FEE, INTERVAL, new SessionDate("2022-01-15", "10:00"));
+        assertThrows(IllegalArgumentException.class, () -> startAfter.lastSessionOnOrBefore(SESSION_DATE));
+    }
+
+
+    @Test
+    void numOfSessionBetweenTest() {
         SessionDate firstSessionDate = new SessionDate("2021-02-28", "10:00");
         SessionDate lastSessionDate = new SessionDate("2021-04-11", "10:00");
         SessionDate firstOfMarch = new SessionDate("2021-03-01", "00:00");
         SessionDate lastOfMarch = new SessionDate("2021-03-31", "23:59");
 
-        int eyeBalledAns1 = 2;
+
         Interval interval1 = new Interval("14");
         //first and last must match
         RecurringSession s = new RecurringSession(firstSessionDate, new Duration("100"), new Subject("Math"),
```

Comment:
Hi maybe this can be session instead of s. Just worried that using s instead would violate the java coding standard.